### ISSUE TITLE
Implement matlab-like cell features

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#852]
_From [ricardob...@gmail.com](https://code.google.com/u/116032747765426583751/) on 2011-11-27T19:14:15Z_

First of all, a big congratulation for making such a nice piece of open software. It's been a month since I've switched to Spyder form Matlab.

Currently I'm finishing my PhD and I'm working with an experimental data set, and after reading and processing millions of data point, I have to plot them y many different ways to search for patterns. 

I very useful feature of Matlab that I miss, is the 'cells'. Thanks to them I could split my code in two big cells, the first for reading and processing my data, and the second to make the plot. Thus, I just ran the first cell one to get processed data. Then, the data was ready to be plotted in many ways that I pictured many times on the fly, without the need to reprocess all my data. A time consuming task.

Would it be possible to incorporate this in seamless way to the python script. Maybe with comment directive, as in Matlab?

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=852_

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2011-11-28T07:05:01Z_

We discussed a lot about this before (see issue #117 ). Maybe it's time to reconsider it because I would also like to have it.

Thanks for your interest in Spyder and we're really glad you like it.
Carlos

**Labels:** -Type-Defect Type-Enhancement Cat-Editor

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2011-11-29T00:39:29Z_

As a non-Matlab user, I'd like to see little demo or screenshot of the idea.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-12-04T05:32:56Z_

Here is an explicit screencast: http://www.youtube.com/watch?v=DiNc-bOqvo8

**Comment by spyder-issues-migrator** [comment]
_From [josephsm...@gmail.com](https://code.google.com/u/107444192537672587895/) on 2011-12-05T07:49:53Z_

As someone who spends half his time using Matlab I can assure you *the* most practical advantage of Matlab's GUI over Spyder is their cell capabilities.    I can understand being skeptical if you have never used cells like a Matlab user but once you start doing it you get hooked over how practical they are.

Please implement similar cell capabilities into Spyder!

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2011-12-06T03:28:37Z_

I'd rather work on some interface to debug stackless threadlets and parallel execution programs. =) So, not being personally interested in this feature, I'd mark this as ContributorNeeded.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2011-12-15T17:37:38Z_

I did some quick tests and I think this is doable without too much modifications to the codebase

In my opinion, we should only use cells to evaluate portions of code on the interactive console. But for this to work flawlessly we need to ensure that the sent code is perfectly indented (else we got indentation errors) and is broken up in well-formed blocks (for example we can't send two def's in one block. It has to partitioned so that we sent one def first and then the other one).

I'll try to work on it to see what we can do about it.

**Summary:** Implement matlab-like cell features

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2011-12-15T17:39:07Z_

issue #117 has been merged into this issue.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2012-03-18T14:12:30Z_

**Labels:** -Type-Enhancement Type-Enh

**Comment by spyder-issues-migrator** [comment]
_From [StatueOf...@gmail.com](https://code.google.com/u/106324076688393526973/) on 2012-03-18T23:01:07Z_

This sounds wonderful. Cell features make debugging and unit testing very fast.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2012-03-18T23:59:54Z_

When a feature sounds interesting for you, please do not hesitate to "star" it as it will increase its position in the Issues list (which is sorted along milestones, then stars rating, then priority).

**Comment by spyder-issues-migrator** [comment]
_From [petrus.h...@gmail.com](https://code.google.com/u/113904071220560210639/) on 2012-05-21T01:51:26Z_

I think this would be a very useful addition for developing scripts such as scientific small programs. I used to use this all the time in matlab, F9 in spyder is not similarly useful as all the lines need to be in a block.

One implementation that seems to work well is the one in idlex, "SubCode" extension, http://idlex.sourceforge.net/extensions.html

**Comment by spyder-issues-migrator** [comment]
_From [petrus.h...@gmail.com](https://code.google.com/u/113904071220560210639/) on 2012-11-17T01:26:02Z_

The Idlex SubCode extension now works with Ipython 0.13. This feature would be really welcomed in Spyder for those used to the Matlab way of script development.

**Comment by spyder-issues-migrator** [comment]
_From [miguelba...@gmail.com](https://code.google.com/u/106675673014048586124/) on 2012-11-21T10:02:18Z_

I switched from MATLAB to Python a few months ago, and I miss the cells. Not only does it allow easily evaluating portions of code, but it also automatically induces you to organize your code better.

BTW, major kudos in making what is, for me, the best Python IDE out there.

**Comment by spyder-issues-migrator** [comment]
_From [ben.ke...@gmail.com](https://code.google.com/u/109100566592319057834/) on 2012-12-03T15:23:53Z_

Note that you can change this yourself locally. Under your python site-packages directory, find the mixins.py file, and specifically the function _is_separator: https://code.google.com/p/spyderlib/source/browse/spyderlib/widgets/sourcecode/mixins.py#186 Note: this was previous in spyderlib/widgets/sourcecode/base.py , so you might find it there.
Change the line return len(text.strip()) == 0 or text.lstrip()[0] == '#' to, for example, return text.strip() == '##'
and tada! Now your code blocks are separated by lines that have consisting only of '##'.
Ideally, this would be a setting ("Use custom block separator"), but until then I have found this solution quite workable.

**Comment by spyder-issues-migrator** [comment]
_From [miguelba...@gmail.com](https://code.google.com/u/106675673014048586124/) on 2012-12-04T01:57:32Z_

Hi Ben,

Thanks for the suggestion! Unfortunately I cannot find that file. I got Spyder through the Ubuntu Software Center. I have version 2.1.9, which is the latest there, unfortunately.

When I type "locate mixins.py" into a terminal the output is

/usr/lib/python2.7/dist-packages/logilab/astng/mixins.py
/usr/lib/python2.7/dist-packages/logilab/astng/mixins.pyc
/usr/lib/python2.7/dist-packages/twisted/internet/test/connectionmixins.py
/usr/lib/python2.7/dist-packages/twisted/internet/test/connectionmixins.pyc
/usr/lib/python2.7/dist-packages/twisted/internet/test/reactormixins.py
/usr/lib/python2.7/dist-packages/twisted/internet/test/reactormixins.pyc
/usr/share/pyshared/logilab/astng/mixins.py
/usr/share/pyshared/twisted/internet/test/connectionmixins.py
/usr/share/pyshared/twisted/internet/test/reactormixins.py

I looked at the first entry and there is no function named "_is_separator". Perhaps v. 2.1.9 didn't have this functionality yet?

**Comment by spyder-issues-migrator** [comment]
_From [ben.ke...@gmail.com](https://code.google.com/u/109100566592319057834/) on 2012-12-04T01:59:25Z_

That's why I said "Note: this was previous in spyderlib/widgets/sourcecode/base.py", though I meant to say "previously". Look there.

**Comment by spyder-issues-migrator** [comment]
_From [miguelba...@gmail.com](https://code.google.com/u/106675673014048586124/) on 2012-12-04T02:56:06Z_

I missed that part. I've found it now, and changed it to return "##" in text, which is roughly what MATLAB used. I can now use F9 to run blocks of code. Thanks! :)

If I may, I'd like to leave here some other MATLAB cell features which I miss, but which should have a low priority:

- Put the first line in each block in bold.
- Separate blocks with horizontal lines (like a \hline in LaTeX).
- Set the background of the block which contains the current cursor position to a slightly different color (for example, a light yellow or gray).

You can see these features at roughly 0:40 in the video linked in post 3. All of these are mere cosmetic features, but I liked the way this helped me visualize the blocks of code at a glance.

Since they are cosmetic only, they should have a low priority...

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2012-12-05T22:06:56Z_

I found the hack ben mentioned a long time ago but after I saw the video, I decided to implement a fuller solution, with the improvements mentioned by miguel.

Unfortunately IPython integration has kept me quite busy, but the good news is that matlab cells are quite easy to create with the new IPython. I'll try to develop them during the next couple of months.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-03-03T16:36:47Z_

**Labels:** MS-v2.3

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-03-03T16:37:38Z_

issue #1292 has been merged into this issue.

**Comment by spyder-issues-migrator** [comment]
_From [sylvain....@gmail.com](https://code.google.com/u/102212202545902308399/) on 2013-03-05T16:42:37Z_

Hi Carlos, 
Even though you plan to implement the more fancy features like coloring and separating with lines in 2.3 or later, wouldn't it be worth to implement to short hack mentioned already? 
The behavior would already be closer to the one that you want to have eventually. 
Sylvain

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-03-06T08:06:38Z_

@-Sylvain: Hacks are hard to maintain, because people can find a lot of corner cases where things don't work as expected.

You can see issue #872 , which came after my two-days hack to automatically add colons and quotes: I haven't been able to make it work right yet :)

**Comment by spyder-issues-migrator** [comment]
_From [ben.ke...@gmail.com](https://code.google.com/u/109100566592319057834/) on 2013-03-06T10:57:01Z_

Also of note is that the change I made doesn't totally suffice: it also needs to remove all blank lines in the block before passing it to the interpreter, since the interpreter sees those as delimiting indented blocks. For example, this will fail with an error about unexpected indentation:

for i in xrange(2):
    print i

    print i+1
print 'done'

**Comment by spyder-issues-migrator** [comment]
_From [sylvain....@gmail.com](https://code.google.com/u/102212202545902308399/) on 2013-03-06T12:56:19Z_

Hi, is it really related to this cell functionality ? 
In the current development version of spyder
   -if you use an ipython console, everything is fine. 
   -if you use a regular console and select the piece of code you just wrote in the editor and run the selection, it fails. In a sense, it is already a bug of the current version and related to the hack in "select_current_block". 

Maybe the selected text should then be cleaned from blank lines afterwards. 
S.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-03-24T05:58:29Z_

issue #117 has been merged into this issue.

**Comment by spyder-issues-migrator** [comment]
_From [ben.ke...@gmail.com](https://code.google.com/u/109100566592319057834/) on 2013-03-25T00:11:46Z_

I like the "#%%" idea. It's unlikely to be encountered otherwise in a script, because the magic functions are only for interactive mode, and %% denotes a cell concept in ipython too.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-03-25T07:03:17Z_

Yep, and it would also let us do things like:

\#%%timeit
for i in li:
   ...

when sending selections to the IPython console, which would be awesome!

**Comment by spyder-issues-migrator** [comment]
_From [xianghu...@gmail.com](https://code.google.com/u/105653643796078715096/) on 2013-05-02T22:08:25Z_

My first day using spyder and my first post.
This is one of the few features I'm missing as a former matlab user. 
Normally how long will it takes to a new feature like this to be implemented in spyder? I hope I could hold my breath for it.

**Comment by spyder-issues-migrator** [comment]
_From [blueo...@gmail.com](https://code.google.com/u/114677998474882230768/) on 2013-06-29T02:53:21Z_

I totally think this is a valuable feature. It's not a Matlab feature (although Matlab does implement it). It's a scientific scripting feature. If you've ever found yourself marking the same text block as selected, again and again, just to run it, then you'll definitely find this feature useful.
I admit at first sight it may seem like a weird feature - why would you want execute a working script in blocks? The thing is - when you think about the way scientific code is written - in iterations - until it 'hopefully' does the right thing - this 'run block' behavior really fits in, and I'm speaking from my own experience.

In short - I really think this is a very valuable feature, and thanks for Spyder - an awesome piece of software!

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2013-07-03T09:15:45Z_

**Labels:** -MS-v2.3 MS-v2.4

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-08-16T13:03:40Z_

This issue was updated by revision 3a60546bfb72 .


The official block separator is now #%%. This can be changed in spyder.ini.
I really don't think that changing this option is required in the Preferences
dialog box. In fact, the best would even be to set this block separator once and
for all, to allow better portability of scripts from a Spyder setup to another.

**Status:** Fixed

**Comment by spyder-issues-migrator** [comment]
_From [ben.ke...@gmail.com](https://code.google.com/u/109100566592319057834/) on 2013-08-16T13:42:35Z_

I think #%% is a good compromise. Can text be added on that line, like "#%% This cell does X"? That seems useful, especially for similarity with matlab.

**Comment by spyder-issues-migrator** [comment]
_From [sylvain....@gmail.com](https://code.google.com/u/102212202545902308399/) on 2013-08-16T13:44:36Z_

Hi Pierre, 

* when exporting an ipython notebook into a script, the separator is "# \<codecell>". It would be great to have this supported by default as well... Hence, we could have a list of supported separators, rather than a single one. 

* if a persons edits the source code with eclipse and uses eclipse's automatic code formating, "#%%..." is transformed into "# %%...". Hence, every separator of the form "#string" in the list should have two versions supported 
     - the original "#string"
     - and "# string" (or any form of "#      string") 

What do you think?

Sylvain

**Comment by spyder-issues-migrator** [comment]
_From [ben.ke...@gmail.com](https://code.google.com/u/109100566592319057834/) on 2013-08-16T14:12:09Z_

Also, it doesn't seem to advance to the next cell when running, which was the behavior with my simple hack. There should probably be both options (run selection, run selection and go to next), like in matlab.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-08-17T01:44:14Z_

This issue was updated by revision a54c37f4b957 .

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-08-17T02:23:04Z_

This issue was updated by revision 248410d793ef .

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-08-17T02:23:40Z_

@-ben: added "Run block and advance" action in "Source" menu with revision a54c37f4b957 .

@-sylvain: added support for "# %%" and "# \<codecell>" block separators with revision 248410d793ef .

**Comment by spyder-issues-migrator** [comment]
_From [ben.ke...@gmail.com](https://code.google.com/u/109100566592319057834/) on 2013-08-17T16:07:47Z_

Awesome!

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-08-18T02:21:54Z_

This issue was updated by revision 5fea82cfe4d2 .

**Comment by spyder-issues-migrator** [comment]
_From [blueo...@gmail.com](https://code.google.com/u/114677998474882230768/) on 2013-08-18T13:53:32Z_

Great work!

**Comment by spyder-issues-migrator** [comment]
_From [sylvain....@gmail.com](https://code.google.com/u/102212202545902308399/) on 2013-08-19T07:57:03Z_

Hello, 
I have the impression that the "run selection" and "run block" should be very different things. The current behavior of the "run selection or current block" context menu is to run the selection if not void and run the "block" if nothing is selected. Then, after execution, the block remains selected. If you use "run block and advance", the next block is selected... 

I would suggest to have the three following context menus, one being.
  * "run selection", which runs the selection and does nothing if nothing is selected. (With F9 keyboard shortcut as in Matlab)
  * "run block", which runs the current block and does not change the position of the cursor and does not select anything. (With Crtl-Enter keyboard shortcut as in Matlab or IPython in-place cell execution)
  * "run block and advance", which runs the current block and puts the cursor at the beginning of the next cell without selecting anything. (With Shift-Enter keyboard shortcut as in IPython)

Another question is why not calling it "cell" instead of "block"?

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-08-19T08:57:34Z_

I completely agree with you, Sylvain.

Except for a detail perhaps: we may not be able to provide the Ctrl+Enter and Shift+Enter shortcuts (I remember having difficulties implementing such a keyboard shortcut other than hard-coding it). But it has to be confirmed.

I'm reopening the issue to take into account your excellent suggestions.

**Status:** Started

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-08-20T05:16:52Z_

This issue was updated by revision aa5e0063eda7 .

**Status:** Fixed

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2013-08-20T06:53:26Z_

**Labels:** -MS-v2.4 MS-v2.2.3

**Comment by spyder-issues-migrator** [comment]
_From [sylvain....@gmail.com](https://code.google.com/u/102212202545902308399/) on 2014-01-13T15:11:46Z_

Hi Guys, it seems that in the next IPython (2.0), the roles of Ctrl+Enter and Shift+Enter will be reverted. 

cf. https://github.com/ipython/ipython/pull/3605

**Comment by spyder-issues-migrator** [comment]
_From [The1Man...@gmail.com](https://code.google.com/u/113077170601917386097/) on 2014-03-02T13:50:24Z_

Hi all,
Like many others I'm toying with switching from Matlab to Python.  I've spent a large fraction of the past week writing code in Spyder and on the whole Python seems to be a lot more pleasant than Matlab, but..and Spyder is an impressive piece of software...but..

I do miss the simplicity of the Matlab debugger and its profiler.  

Also, the way Spyder does *cells* isn't massively useful as it currently stands (in 2.2.5, and as far as I can tell from the feature list for future releases):

* It would be nice if running a cell invokes the UMD, or at least there was a version of the #%% syntax which does, e.g. #%%-- or something.
*  Why does the code have to be printed line by line to the console? It dilutes my command history, scrolls the console, and it doesn't seem very quick.
*  Also, and this is related to the previous point, breakpoints/ explicit pdb stack_trace()s don't work in cell mode (incidentally Matlab doesn't let you set break points in cell mode either, but I don't see why it could be implemented in principle).

My guess, without looking at the source myself, is that all of these requests could be relatively simply implemented by creating a temporary file cell.py (or whatever) and then running that in the console.

I appreciate that even simple things take time and effort to get right..I would offer to help myself but that may not be a great idea given I'm still fairly new to Python (though already pretty obsessed it seems).

Fingers crossed for more work on the debugging UI and thanks again for all the great work!

**Comment by spyder-issues-migrator** [comment]
_From [contrebasse](https://code.google.com/u/contrebasse/) on 2014-03-04T06:03:15Z_

If you want a line-by-line profiler similar to matlab (but not as complete, matlab's one is really great), you can use http://pythonhosted.org/line_profiler/ I wrote a plugin to integrate it in spyder : https://github.com/Nodd/spyder_line_profiler Note that it profiles only the functions that have a @-profile decorator, so it won't work if you only use a big script with cells.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2015-01-02T17:43:06Z_

issue #925 has been merged into this issue.
